File: Algorithm_I/3_Longest_Substring_Without_Repeating_Characters_a6.cc
#include <iostream>
#include <vector>


// Time to complete: 3.5 hrs

// Needed to research:
//   - std::map methods

//// Begin solution, results at https://leetcode.com/submissions/detail/740530934/


#include <map>
#include <algorithm>


class Solution {
public:
    // 0 <= s.length <= 5 * 10^4
    // s consists of English letters, digits, symbols and spaces
    // Given a string s, find the length of the longest substring without
    //   repeating characters.
    int lengthOfLongestSubstring(const std::string &s) {
        const int s_sz (s.size());
        if (s_sz < 2)
            return s_sz;
        int max_ss_sz {0};  // max substring length
        std::map<char, int> ss_indices;  // substring char: index in s
        for (int i {0}, ss_sz {0}; i < s_sz; ++i, ss_sz = 0) {
            ss_indices[s[i]] = i;  // update if present, insert if not
            for (int j {i}; j >= 0 && j == ss_indices[s[j]]; --j)
                ++ss_sz;
            max_ss_sz = std::max(max_ss_sz, ss_sz);
        }
        return max_ss_sz;
    };
};

//// end solution


int main() {
    const std::string s1 {"abcabcbb"};
    const std::string s2 {"bbbbb"};
    const std::string s3 {"pwwkew"};
    const std::string s4 {" "};
    const std::string s5 {"dvdf"};
    const std::string s6 {"asjrgapa"};
    const std::string s7 {"aabaab!bb"};
    const std::string s8 {"uqinntq"};
    const std::string s9 {"au"};

    Solution solution;

    std::cout << "solution.lengthOfLongestSubstring(\"" << s1 << "\"): " <<
        solution.lengthOfLongestSubstring(s1) << '\n';
    // 3 ("abc")

    std::cout << "solution.lengthOfLongestSubstring(\"" << s2 << "\"): " <<
        solution.lengthOfLongestSubstring(s2) << '\n';
    // 1 ("b")

    std::cout << "solution.lengthOfLongestSubstring(\"" << s3 << "\"): " <<
        solution.lengthOfLongestSubstring(s3) << '\n';
    // 3 ("wke")

    std::cout << "solution.lengthOfLongestSubstring(\"" << s4 << "\"): " <<
        solution.lengthOfLongestSubstring(s4) << '\n';
    // 1 (" ")

    std::cout << "solution.lengthOfLongestSubstring(\"" << s5 << "\"): " <<
        solution.lengthOfLongestSubstring(s5) << '\n';
    // 3 ("vdf")

    std::cout << "solution.lengthOfLongestSubstring(\"" << s6 << "\"): " <<
        solution.lengthOfLongestSubstring(s6) << '\n';
    // 6 ("sjrgap")

    std::cout << "solution.lengthOfLongestSubstring(\"" << s7 << "\"): " <<
        solution.lengthOfLongestSubstring(s7) << '\n';
    // 3 ("ab!")

    std::cout << "solution.lengthOfLongestSubstring(\"" << s8 << "\"): " <<
        solution.lengthOfLongestSubstring(s8) << '\n';
    // 4 ("uqin")

    std::cout << "solution.lengthOfLongestSubstring(\"" << s9 << "\"): " <<
        solution.lengthOfLongestSubstring(s9) << '\n';
    // 2 ("au")
}

File: Algorithm_I/3_Longest_Substring_Without_Repeating_Characters_a6.py
#!/usr/bin/python3


# Time to complete: 3.5 hrs

# Needed to research:

#### Begin solution, results at https://leetcode.com/submissions/detail/740528683/


class Solution:
    """
    """
    # 0 <= s.length <= 5 * 10^4
    # s consists of English letters, digits, symbols and spaces
    def lengthOfLongestSubstring(self, s: str) -> int:
        """Finds the length of the longest substring without repeating
        characters.
        """
        if len(s) < 2:
            return len(s)
        ss_indices = {}  # substring char: index in s
        max_ss_len = 0
        for i, c in enumerate(s):
            ss_indices[c] = i
            ss_len = 0
            for j in range(i, -1, -1):
                if j != ss_indices[s[j]]:
                    break
                ss_len += 1
            max_ss_len = max(max_ss_len, ss_len)
        return max_ss_len


#### end solution


if __name__ == '__main__':
    s1 = "abcabcbb"
    s2 = "bbbbb"
    s3 = "pwwkew"
    s4 = " "
    s5 = "dvdf"
    s6 = "asjrgapa"
    s7 = "aabaab!bb"
    s8 = "uqinntq"
    s9 = "au"

    solution = Solution()

    print(f'solution.lengthOfLongestSubstring("{s1}"):',
          f'{solution.lengthOfLongestSubstring(s1)}')
    # 3 ("abc")

    print(f'solution.lengthOfLongestSubstring("{s2}"):',
          f'{solution.lengthOfLongestSubstring(s2)}')
    # 1 ("b")

    print(f'solution.lengthOfLongestSubstring("{s3}"):',
          f'{solution.lengthOfLongestSubstring(s3)}')
    # 3 ("wke")

    print(f'solution.lengthOfLongestSubstring("{s4}"):',
          f'{solution.lengthOfLongestSubstring(s4)}')
    # 1 (" ")

    print(f'solution.lengthOfLongestSubstring("{s5}"):',
          f'{solution.lengthOfLongestSubstring(s5)}')
    # 3 ("vdf")

    print(f'solution.lengthOfLongestSubstring("{s6}"):',
          f'{solution.lengthOfLongestSubstring(s6)}')
    # 6 ("sjrgap")


    print(f'solution.lengthOfLongestSubstring("{s7}"):',
          f'{solution.lengthOfLongestSubstring(s7)}')
    # 3 ("ab!")

    print(f'solution.lengthOfLongestSubstring("{s8}"):',
          f'{solution.lengthOfLongestSubstring(s8)}')
    # 4 ("uqin")

    print(f'solution.lengthOfLongestSubstring("{s9}"):',
          f'{solution.lengthOfLongestSubstring(s9)}')
    # 2 ("au")
